File: javascript/primes/Atkin.js
#!/usr/bin/env node

'use strict';

const UPPER_BOUND = 5000000;
const PREFIX = 32338;

class Node {
    constructor() {
        this.children = new Map();
        this.terminal = false;
    }
}

const toInt = function(n) {
    return n >>> 0;
}

class Sieve {
    constructor(limit) {
        this.limit = limit;
        this.prime = new Int8Array(limit + 1);
    }

    toList() {
        const result = [2, 3];
        for (let p = 5; p <= this.limit; ++p) {
            if (this.prime[p]) {
                result.push(p);
            }
        }
        return result;
    }

    omitSquares() {
        for (let r = 5; r * r < this.limit; ++r) {
            if (this.prime[r]) {
                for (let i = r * r; i < this.limit; i += r * r) {
                    this.prime[i] = 0;
                }
            }
        }
        return this;
    }

    step1(x, y) {
        const n = (4 * x * x) + (y * y);
        if (n <= this.limit && (n % 12 == 1 || n % 12 == 5)) {
            this.prime[n] = !this.prime[n];
        }
    }

    step2(x, y) {
        const n = (3 * x * x) + (y * y);
        if (n <= this.limit && n % 12 == 7) {
            this.prime[n] = !this.prime[n];
        }
    }

    step3(x, y) {
        const n = (3 * x * x) - (y * y);
        if (x > y && n <= this.limit && n % 12 == 11) {
            this.prime[n] = !this.prime[n];
        }
    }

    loopY(x) {
        for (let y = 1; y * y < this.limit; ++y) {
            this.step1(x, y);
            this.step2(x, y);
            this.step3(x, y);
        }
    }

    loopX() {
        for (let x = 1; x * x < this.limit; ++x) {
            this.loopY(x);
        }
    }

    calc() {
        this.loopX();
        return this.omitSquares();
    }
}

function generateTree(l) {
    const root = new Node()
    for (const el of l) {
        let head = root;
        for (const ch of el.toString()) {
            if (!head.children.has(ch)) {
                head.children.set(ch, new Node());
            }
            head = head.children.get(ch);
        }
        head.terminal = true;
    }
    return root;
}

function find(upperBound, prefix) {
    const primes = new Sieve(upperBound).calc();
    const strPrefix = prefix.toString();
    let head = generateTree(primes.toList());

    for (const ch of strPrefix) {
        head = head.children.get(ch);
        if (typeof head === 'undefined') {
            return null;
        }
    }

    let [queue, result] = [[[head, strPrefix]], []]
    while (queue.length > 0) {
        const [top, prefix] = queue.pop();
        if (top.terminal) {
            result.push(toInt(prefix));
        }

        for (const [ch, v] of top.children) {
            queue.splice(0, 0, [v, prefix + ch]);
        }
    }
    return result;
}

(async function() {
    const startTimeMs = new Date().getTime();
      
    console.log(find(UPPER_BOUND, PREFIX));

    const endTimeMs = new Date().getTime();
    const durationMs = endTimeMs - startTimeMs;
    console.log(`Execution time: ${durationMs}ms`);
})();

File: c-plus-plus/primes/Atkin.cpp
#include <iostream>
#include <memory>
#include <queue>
#include <sstream>
#include <unistd.h>
#include <map>
#include <utility>
#include <vector>
#include <chrono>

static const auto UPPER_BOUND = 5'000'000;
static const auto PREFIX = 32'338;

struct Node {
  std::map<char, std::shared_ptr<Node>> children{};
  bool terminal = false;
};

class Sieve {
  int limit;
  std::vector<bool> prime;

  Sieve& omit_squares() {
    for (auto r = 5; r * r < limit; ++r) {
      if (prime[r]) {
        for (auto i = r * r; i < limit; i += r * r) {
          prime[i] = false;
        }
      }
    }
    return *this;
  }

  void step1(int x, int y) {
    const auto n = (4 * x * x) + (y * y);
    if (n <= limit && (n % 12 == 1 || n % 12 == 5)) {
      prime[n] = !prime[n];
    }
  }

  void step2(int x, int y) {
    const auto n = (3 * x * x) + (y * y);
    if (n <= limit && n % 12 == 7) {
      prime[n] = !prime[n];
    }
  }

  void step3(int x, int y) {
    const auto n = (3 * x * x) - (y * y);
    if (x > y && n <= limit && n % 12 == 11) {
      prime[n] = !prime[n];
    }
  }

  void loop_y(int x) {
    for (auto y = 1; y * y < limit; ++y) {
      step1(x, y);
      step2(x, y);
      step3(x, y);
    }
  }

  void loop_x() {
    for (auto x = 1; x * x < limit; ++x) {
      loop_y(x);
    }
  }

public:
  Sieve(int limit): limit(limit), prime(limit + 1, false) {}

  std::vector<int> to_list() const {
    std::vector<int> result({2, 3});
    for (auto p = 5; p <= limit; ++p) {
      if (prime[p]) {
        result.push_back(p);
      }
    }
    return result;
  }

  Sieve& calc() {
    loop_x();
    return omit_squares();
  }
};

std::shared_ptr<Node> generate_tree(const std::vector<int>& l) {
  auto root = std::make_shared<Node>();
  for (const auto el : l) {
    auto head = root;
    for (const auto ch: std::to_string(el)) {
      head->children.emplace(ch, std::make_shared<Node>());
      head = head->children[ch];
    }
    head->terminal = true;
  }
  return root;
}

std::vector<int> find(int upper_bound, int prefix) {
  const auto primes = Sieve(upper_bound).calc();
  const auto& str_prefix = std::to_string(prefix);
  auto head = generate_tree(primes.to_list());

  for (const auto ch : str_prefix) {
    head = head->children[ch];
  }

  std::queue<std::pair<std::shared_ptr<Node>, std::string>> queue(
    { std::make_pair(head, str_prefix) }
  );
  std::vector<int> result;
  while (!queue.empty()) {
    const auto& [top, prefix] = queue.front();
    queue.pop();
    if (top->terminal) {
      result.push_back(std::stoi(prefix));
    }
    for (const auto& [ch, v] : top->children) {
      queue.push(std::make_pair(v, prefix + ch));
    }
  }
  return result;
}

std::string to_string(const std::vector<int>& a) {
  std::stringstream ss;
  ss << "[";
  auto first = true;
  for (const auto el : a) {
    if (!first) {
      ss << ", ";
    }
    ss << el;
    first = false;
  }
  ss << "]";
  return ss.str();
}

int main() {
  const auto start_time = std::chrono::high_resolution_clock::now();

  std::cout << to_string(find(UPPER_BOUND, PREFIX)) << std::endl;
  
  const auto end_time = std::chrono::high_resolution_clock::now();
  const auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time).count();

  std::cout << "Execution time: " << duration << "ms" << std::endl;
}